core/src/main/java/me/aurium/beetle/api/task/futures/TaskFuture.java

```diff
@@ -0,0 +1,230 @@
+package me.aurium.beetle.api.task.futures;
+
+import me.aurium.beetle.api.task.ExecutorProvider;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.*;
+
+/**
+ * Represents a future that is typically backed by a common async executor (typically a tasker's) rather than from the common pool.
+ *
+ * Still allows access to specified executors rather than the integrated executor.
+ *
+ * @param <T>
+ */
+@SuppressWarnings("unchecked")
+public class TaskFuture<T> extends CompletableFuture<T> {
+
+    private final Executor asyncExecutor;
+
+    public TaskFuture(ExecutorProvider executorProvider) {
+        this.asyncExecutor = executorProvider.getAsyncExecutor();
+    }
+
+    @Override
+    public Executor defaultExecutor() {
+        return asyncExecutor;
+    }
+
+    @Override
+    public <U> TaskFuture<U> thenApply(Function<? super T, ? extends U> fn) {
```

Comment:
Taskfutures are guarunteed to use the async executor of a tasker as well as provide the tasker that they used (i have to add that back)

Reply:
I understand that. Does knowing that you have a TaskFuture provide any value over knowing that you have a CompletableFuture? I doubt it. It is within the contract of CompletableFuture that the default executor may be overridden.

Reply:
No, but having access to the tasker (read: not the executorprovider) is beneficial?

Reply:
TaskFuture doesn't provide access to the tasker, even if it did, I don't think that would be helpful and seems more like exposing an implementation detail.

Instead of referring to objects as TaskFuture, I suggest referring to them as CompletableFuture, because the API is the same and TaskFuture provides no distinction other than obscurity.